File: Design and Analysis of Algorithms Summary/CA2/q1/q1.py
import sys
sys.setrecursionlimit(100000)

mem = 100000 * [-1]

def f(mem, k, space):
    if mem[space] != -1:
        return mem[space]
    if space < k:
        return 1
    if space == k:
        return 2
    if space > k:
        ans = f(mem, k, space-k) + f(mem, k, space-1)
        mem[space] = ans
        return ans

if __name__ == '__main__':
    n, k = map(int, input().split(' '))
    for _ in range(n):
        start, end = map(int, input().split(' '))
        res = 0
        for i in range(start, end+1):
            res += f(mem, k, i)
        print(res % 1000000007)

File: Design and Analysis of Algorithms Summary/CA2/q1/q1.cpp
#include <iostream>
using namespace std;

#define MAX_N 100000
#define MOD 1000000007


int k;
int mem[MAX_N+5];
int ans[MAX_N+5];

int f(int space) {
    if(mem[space] != 0)
        return mem[space];

    if(space < k)
        return 1;

    if(space == k)
        return 2;

    if(space > k) {
        int a = (f(space-k) + f(space-1)) % MOD;
        mem[space] = a;
        return a;
    }
    return -1;
}

int find_ans(int space) {
    if(space == 0)
        return 0;

    if(ans[space] != 0)
        return ans[space];

    int a = (find_ans(space-1) + f(space)) % MOD;
    ans[space] = a;
    return a;
}

int main() {
    int n;
    cin >> n >> k;

    for(int i = 0; i < n; i++) {
        int start, end;
        cin >> start >> end;

        cout << (find_ans(end) - find_ans(start-1) + MOD) % MOD << endl;
    }
}
